app/scripts/models/reliers/oauth.js

```diff
@@ -61,6 +66,83 @@ define([
       return true;
     },
 
+    /**
+     * OAuth reliers can opt in to fetch relier-specific keys.
+     */
+    isKeyFetchEnabled: function () {
+      if (this.get('keys')) {
+        return true;
+      }
+      return Relier.prototype.isKeyFetchEnabled.call(this);
+    },
+
+    deriveRelierKeys: function (uid, keys) {
+      var self = this;
+      var relierKeys = {};
+      return this._generateDerivedKA(uid, keys.kA)
+        .then(function (kAr) {
+          relierKeys.kAr = kAr;
+          return self._generateDerivedKB(uid, keys.kB);
+        }).then(function (kBr) {
+          relierKeys.kBr = kBr;
+          return relierKeys;
+        });
+    },
+
+    _generateDerivedKA: function (uid, kA) {
+      var info = Constants.OAUTH_DERIVED_KEY_INFO_PREFIX;
+      info += 'kAr:' + this.get('clientId');
+      return this._generateDerivedKey(uid, kA, info)
+        .then(function (kAr) {
+          // Indicate key type in kid as an extra debugging aid.
+          kAr.kid = 'kAr-' + kAr.kid;
+          return kAr;
+        });
+    },
+
+    _generateDerivedKB: function (uid, kB) {
+      var info = Constants.OAUTH_DERIVED_KEY_INFO_PREFIX;
+      info += 'kBr:' + this.get('clientId');
+      return this._generateDerivedKey(uid, kB, info)
+        .then(function (kBr) {
+          // Indicate key type in kid as an extra debugging aid.
+          kBr.kid = 'kBr-' + kBr.kid;
+          return kBr;
+        });
+    },
+
+    _generateDerivedKey: function (uid, key, info) {
+      var self = this;
+      key = sjcl.codec.hex.toBits(key);
+      info = sjcl.codec.utf8String.toBits(info);
+      var salt = sjcl.codec.hex.toBits('');
+      // HKDF-derive twice the desired key length.  We'll use the
+      // first chunk for an opaque key id, the second chunk for the actual key.
+      return hkdf(key, info, salt, Constants.OAUTH_DERIVED_KEY_LENGTH * 2)
+        .then(function (out) {
+          // The first 32 bytes form the key id.
+          // The second 32 bytes are the actual key material.
+          var keySizeBits = Constants.OAUTH_DERIVED_KEY_LENGTH * 8;
+          var keyId = sjcl.bitArray.bitSlice(out, 0, keySizeBits);
+          var keyBytes = sjcl.bitArray.bitSlice(out, keySizeBits, keySizeBits * 2);
+          return {
+            kid: self._base64UrlEncode(keyId),
+            k: self._base64UrlEncode(keyBytes),
+            kty: 'oct',
+            rid: self.get('clientId'),
+            uid: uid
```

Comment:
FYI the names `rid` and `uid` here were chosen to mirror the proposed terminolofy in https://github.com/mozilla/fxa-notification-server/blob/master/docs/overview.md#core-concepts